### PR TITLE
Fix: make `ggml_backend_qnn_buffer_type_context` as static also

### PR DESCRIPTION
**Pull request by chraac**
https://github.com/ggerganov/llama.cpp/pull/6869#discussion_r1631100881

From the discussion here, thought we should make the lifespan of type's context match the type iself.

### CODE REVIEW

**Comment by zhouwg** [comment]
you are right. can you submit a new PR? because there are merge conflicts.thanks

File: ggml-qnn.cpp [hunk 1]
```diff
@@ -3090,7 +3093,7 @@ ggml_backend_buffer_type_t ggml_backend_qnn_buffer_type(size_t device) {
                     /* .supports_backend = */ ggml_backend_qnn_buffer_type_supports_backend,
                     /* .is_host          = */ ggml_backend_qnn_buffer_is_host
                 },
-                /* .context = */ new ggml_backend_qnn_buffer_type_context { device, GGML_QNN_NAME + std::to_string(device) },
+                /* .context = */ &context,
```
Comment (hmartinez82) [inline ggml-qnn.cpp:3096]: omg yes, thank you :)

Comment (zhouwg) [reply ggml-qnn.cpp:3096]: this modification is not make sense.

**Comment by zhouwg** [comment]
this memory leak was introduced by your review suggestion and my corresponding modifications(which follow [the style in Intel SYCL backend](https://github.com/ggerganov/llama.cpp/blob/master/ggml-sycl.cpp#L16608) and [CUDA backend from the original author of ggml backend subsysem](https://github.com/ggerganov/llama.cpp/blob/master/ggml-cuda.cu#L625)). btw, my previous code has no such memory leak issue because ctx in my original code here is NULL which follow the style in metal backend(from the original author of this great project. as I said in your PR in my forked project llama.cpp: there are so many so-called language master, but there are only one original author of this real excellent machine learning framework in our planet although there are many machine learning frameworks in the IT industry):
Screenshot from 2024-06-08 23-15-41

**Comment by zhouwg** [comment]
thanks for your PR and time. It has been imported to this branch but not be merged because this is a long-term issue in upstream llama.cpp.

I'd like to close this PR accordingly of course you can re-open it as your need.

thanks so much.

**Pull request closed without merging**